### ISSUE TITLE
RStudio and R crashes with fatal error linked to data.table operation

### ISSUE THREAD

**Issue by AmyMikhail** [Rdatatable/data.table#2672]
I am attempting to summarize some variables cumulatively by group for each week in which there was new activity in that group, with a data.table line listing as input.  This process works fine with a toy version of the function and a small input data set; however with larger data sets and the real (longer) function, R crashes with a fatal error.  The details of the crash are here:

```
Problem signature:
Problem Event Name:	APPCRASH
Application Name:	rsession.exe
Application Version:	1.1.383.0
Application Timestamp:	59d5818a
Fault Module Name:	datatable.dll
Fault Module Version:	0.0.0.0
Fault Module Timestamp:	5a39aedc
Exception Code:	c0000005
Exception Offset:	0000000000029060
OS Version:	6.1.7601.2.1.0.256.48
Locale ID:	2057
Additional Information 1:	0af1
Additional Information 2:	0af12ed5ffbb0678966a6b7fe308a74b
Additional Information 3:	2514
Additional Information 4:	251401539b18800e11472f6b420048c8`
```

Session info is here:

```
> sessionInfo()
R version 3.4.2 (2017-09-28)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252    LC_MONETARY=English_United Kingdom.1252
[4] LC_NUMERIC=C                            LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] rgisws_0.1.2        devtools_1.13.4     openxlsx_4.0.17     scales_0.5.0        ggplot2_2.2.1       stringdist_0.9.4.6 
 [7] stringr_1.3.0       stringi_1.1.6       ISOweek_0.6-2       lubridate_1.6.0     data.table_1.10.4-3 PKI_0.1-5.1        
[13] base64enc_0.1-3     digest_0.6.13       getPass_0.2-2       RPostgreSQL_0.6-2   DBI_0.7            

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.14     magrittr_1.5     munsell_0.4.3    colorspace_1.3-2 rlang_0.1.6      plyr_1.8.4       tools_3.4.2     
 [8] parallel_3.4.2   grid_3.4.2       gtable_0.2.0     withr_2.1.1.9000 yaml_2.1.14      lazyeval_0.2.1   tibble_1.4.2    
[15] memoise_1.1.0    pillar_1.1.0     compiler_3.4.2
```

Here is some example data:
```
# Example data:
mydt <- data.table(id = c("a", "a", "b", "b", "b", "c", "c", "c", "c", "d", "d", "d"),
                   grp = c("G1", "G1", "G1", "G1", "G1", "G2", "G2", "G2", "G2", "G2", "G2", "G2"),
                   name = c("Jack", "John", "Jill", "Joe", "Jim", "Julia", "Simran", "Delia", "Aurora", "Daniele", "Joan", "Mary"),
                   sex = c(NA, NA, "f", "m", "m", "f", "m", "f", "f", "f", "f", "f"), 
                   age = c(2,12,29,15,30,75,5,4,7,55,43,39), 
                   reportweek = c("201740", "201750", "201801", "201801", "201801", "201748", "201748", "201749", "201810", "201752", "201752", "201801"))

# Set levels of group ID to loop over:
idlist <- c("id", "grp")
```

Here is a toy version of the function:

```
sumclusters <- function(dataset, groupvector) {
  
  # Load required packages:
  require(data.table)
  require(lubridate)
  
  #browser()
  
  # Create empty list to hold the results:
  datalist = list()
  
  for (i in unique(dataset$reportweek)) {
    # Subset by cumulative report week
    dpart = subset(dataset, reportweek <= i)
    
    # Calculate summary values within each subset:
    
    # Cluster size, age, duration:
    dpart[, clustersize := .N, by = eval(groupvector)]
    dpart[, maxweek := max(reportweek, 0, na.rm = TRUE), by = eval(groupvector)]
    dpart[, minweek := min(reportweek, na.rm = TRUE), by = eval(groupvector)]
    dpart[, repdate := as.Date(paste(substring(reportweek, 1, 4), substring(reportweek, 5, 6), 1, sep = "-"), "%Y-%U-%u")]
    dpart[, maxwkdate := as.Date(paste(substring(maxweek, 1, 4), substring(maxweek, 5, 6), 1, sep = "-"), "%Y-%U-%u")]
    dpart[, firstdate := min(repdate, na.rm = TRUE), by = eval(groupvector)]
    dpart[, lastdate := max(repdate, na.rm = TRUE), by = eval(groupvector)]
    dpart[, clusteragemonths := (as.numeric(lastdate - firstdate))/(365.25/12)]
    
    # Cluster characteristics last month:
    dpart[, clustersize4wk := sum(reportweek >= isoyrwk(maxwkdate - weeks(4))), by = eval(groupvector)]
    dpart[, fourweekgr := round(clustersize4wk/4, 2)]

    # Age range:
    dpart[, agemin := min(age, na.rm = TRUE), by = eval(groupvector)]
    dpart[, agemax := max(age, 0, na.rm = TRUE), by = eval(groupvector)]
    dpart[, agemedian := round(median(age, na.rm = TRUE), 0), by = eval(groupvector)]
    dpart[, agerange := paste(agemin, " - ", agemax, sep = "")]
    dpart[, adult := length(which(age >= 16)), by = eval(groupvector)]
    dpart[, total4age := length(which(!is.na(age))), by = eval(groupvector)]
    dpart[, adultprop := round((adult/total4age)*100, 0)]
    
    # Sex:
    dpart[, female := sum(sex == "f"), by = eval(groupvector)]
    dpart[, male := sum(sex == "m"), by = eval(groupvector)]
    dpart[, total4sex := length(which(!is.na(sex))), by = eval(groupvector)]
    dpart[, maleprop := round((male/total4sex)*100, 0)]
    dpart[, sexratio := paste("F ", female, " : ", "M ", male, sep = "")]
    
    # Add to list:
    datalist[[i]] = dpart
    
  }
  
  # Bind all the summaries together:
  clusterlog = rbindlist(l = datalist, use.names = T, fill = T, idcol = "ReportNo")
  
  # Create list of report columns:
  reportcols = c(eval(groupvector), "clustersize", "minweek", "maxweek", "clusteragemonths", "clustersize4wk", "fourweekgr",
                 "agemin", "agemax", "agemedian", "agerange", "adultprop", 
                 "maleprop", "sexratio")
  
  # Create cluster tab:
  clustertab = clusterlog[, reportcols, with = F]
  
  # Set the key prior to deduplicating:
  setkeyv(clustertab, c(groupvector))
  
  # Deduplicate to get the summary table:
  clustertab = unique(clustertab)
  
  # Sort table by ID and report week:
  setorderv(clustertab, c(groupvector, "maxweek"))
  
  # Create list of line list columns:
  linelistcols = c("id", "grp", "name", "sex", "age", "reportweek")
  
  # Create cluster line list:
  clusterll = dataset[, linelistcols, with = F]
  
  # Return the summarised data.tables in a list:
  dtlist <- list("linelist" = clusterll, "clustersum" = clustertab)
  return(dtlist)
  
}
```

And dependent function isoyrwk:
```
isoyrwk <- function(dates){
  require(lubridate)
  require(data.table)
  
  isoweekcalc = data.table(myisoweek = isoweek(dates),
                            myisoyear = isoyear(dates),
                            calcweek = week(dates),
                            calcyear = lubridate::year(dates))
  isoweekcalc[ myisoweek >= 52 & calcweek == 1, myisoyear := calcyear]
  isoweekcalc[ myisoweek >= 52 & calcweek == 1, myisoweek := 1]
  isoweekcalc[ myisoweek >= 52 & calcweek >= 52, myisoyear := calcyear]
  isoweekcalc[ myisoweek >= 52 & calcweek >= 52, myisoweek := 52]
  isoweekcalc[, myisoweek := as.character(myisoweek)]
  isoweekcalc[ nchar(myisoweek) == 1, myisoweek := paste0("0", myisoweek)]
  isoweekcalc[, isoyearweek := paste0(myisoyear, myisoweek)]
  isoyearweek = isoweekcalc$isoyearweek
  isoyearweek
}

```
This is how I am applying the sumclusters function to my data:
`mytest <- sapply(idlist, sumclusters, data = mydt, simplify = FALSE, USE.NAMES = TRUE)`

Unfortunately I am not able to reproduce the fatal error with the toy data set and toy function.  The only difference between the toy function and the real one is that there are more conditional counts on different variables, but the strategy for each one is exactly the same as shown above.  I was originally getting a RHS / LHS class discrepancy error but this was discussed and resolved in [this](https://stackoverflow.com/questions/49201334/r-data-table-avoid-class-discrepancy-between-rhs-and-lhs?noredirect=1#comment85414672_49201334) Stack Overflow post.

My real input data set is relatively large (2103 rows in the line listing, with reports in 155 weeks and four grouping vectors containing 1271, 144, 108 and 94 groups each, respectively).

I think the error might be due to the function timing out because there is too much data (I have 16GB ram and my .Rproj file is on a network drive) but is there any way to confirm this from the above error?  Or is it a bug?
 
Any insights into why this is causing R to crash and how I could prevent this would be much appreciated - hope I have posted this in the right place as the error details did specify that the fault module name is `datatable.dll`

**Comment by mrmanojrai** [comment]
Couple of queries/observations:
1. Do you get crash with sample data provided with this issue? 
2. Any specific reason to use `as.Date` when `lubridate` has already been loaded and used?
3. Most of the `j` sections operations could have been grouped and executed in one go as `by` argument is same as  `(by = eval(groupvector))`.

**Comment by MichaelChirico** [comment]
have you reproduced this on the command line?

**Comment by AmyMikhail** [comment]
@mrmanojrai in response to your questions:

1. Unfortunately I have not been able to create a small data set that reproduces the error.  I'm pretty sure there are no operations in my real function that are not represented in the toy one above, but just in case I missed anything please find below the longer function I am using on my actual data.  I suspect that it is some combination of missing values with the group and week selections, but can't think what as I made one group NA for sex in the example above and it still doesn't replicate the crash.

2. This is not really relevant to the problem at hand (at least I don't think it is), but I needed to get a date from the year week as my ISO year weeks (created with the function which handles year-end dates differently to the equivalent function in lubridate) are stored as numeric and I can't perform mathematical operations directly on them.  If there is a way to do this with lubridate I would be happy to change this but I'm not aware of a `yw` function in lubridate? Essentially I was looking for a way to iteratively define four week periods based on the maximum week by group for each iteration - 4 weeks.

3. Could you elaborate on this with an example?

@MichaelChirico not sure what you mean viz. command line - do you mean what happens if I run this in base R rather than RStudio?  I will try this and let you know the outcome.

The longer function is here (note that I have been running this function without the cumulative iteration by week (i.e. no for loop, no `rbindlist`) without problems).  So the introduction of the for loop and some small adjustments I made to get the cumulative summaries by week seem to be causing the crash:

```
#' @title  CLUSTER SUMMARY FUNCTION - AUDIT REPORT VERSION
#' 
#' @author Amy Mikhail (Amy.Mikhail@@phe.gov.uk)
#' 
#' @details 
#' This function generates the cluster summaries for audits (start of routine WGS to date).
#' The cluster summary output is presented in a weekly log format.
#' In this way changes in cluster designation over time can be monitored.
#' 
#' @param linelist Line list for SNP clusters to be summarised (note: must be geocoded already)
#' @param groupvector Character vector of group IDs (SNP levels) to summarise over
#' 
#' @return List of two objects: (a) clean line list and (b) Cluster summary log by week for each SNP level
#' 
#' @export
sumclustersaudit <- function(linelist, groupvector) {
  
  # Load required libraries:
  require(data.table)
  require(lubridate)
  
  #browser()
  
  # Create empty list to hold the results:
  datalist = list()
  
  for (i in unique(linelist$reportweek)) {
    
    # Subset by cumulative report week
    dpart = subset(linelist, reportweek <= i)
    
    # Define iteration-specific SNP address:
    dpart[, xsnpaddress := ifelse(eval(groupvector) == "groupid25", snpaddress_t25, 
                                         ifelse(eval(groupvector) == "groupid10", snpaddress_t10, 
                                                ifelse(eval(groupvector) == "groupid05", snpaddress_t5, snip_address))), by = 1:nrow(dpart)]
    
    # Calculate number of human cases in each 5-SNP cluster:
    dpart[, clustersize := sum(sampleorigin == "human"), by = eval(groupvector)]
    
    # Calculate number of isolates (including non-human) in each cluster:
    dpart[, totalisolates := .N, by = eval(groupvector)]
    
    # Identify if there are food, animal or environmental sample types in each cluster:
    dpart[, linkedsamples := paste0(unique(sampleorigin[sampleorigin != "human"]), collapse = " "), by = eval(groupvector)]
    #dpart[ linkedsamples %in% c("", " "), linkedsamples := NA]
    
    # Remove SNP addresses comprising a single case (orphans after deduplication):
    dpart = dpart[ totalisolates > 1 ]
    
    # Identify if there are human systemic samples in each cluster (marker for invasive disease):
    dpart[, invasivecases := sum(sampleorigin == "human" & invasive == TRUE), by = eval(groupvector)]
    
    # Determine the week and year for the specimen report date & SNP address report date of each specimen:
    dpart[, maxseqweek := max(isorepweekseq, 0, na.rm = TRUE), by = eval(groupvector)]
    dpart[, maxptweek := max(isorepweekpt, 0, na.rm = TRUE), by = eval(groupvector)]
    
    # Determine the week in which the first and last specimens in each cluster were detected:
    dpart[, maxweek := max(reportweek, 0, na.rm = TRUE), by = eval(groupvector)]
    dpart[, minweek := min(reportweek, na.rm = TRUE), by = eval(groupvector)]
    dpart[ sampleorigin == "human", minweekhuman := min(reportweek, 0, na.rm = TRUE), by = eval(groupvector)]
    dpart[, firstdate := min(specimendate, na.rm = TRUE), by = eval(groupvector)]
    dpart[, lastspecdate := max(specimendate, na.rm = TRUE), by = eval(groupvector)]
    dpart[, lastseqdate := max(reportdate_seq, na.rm = TRUE), by = eval(groupvector)]
    
    # Determine the number of weeks each cluster was active:
    dpart[, weeksactive := length(unique(reportweek)), by = eval(groupvector)]
    
    # Determine cluster age in months and growth rate according to specimen date:
    dpart[, clusteragemonths := (as.numeric(lastspecdate - firstdate))/(365.25/12)]
    dpart[ clusteragemonths == 0, clusteragemonths := 0.03]
    dpart[ clusteragemonths >= 1, clusteragemonths := round(clusteragemonths, 0)]
    dpart[ clusteragemonths < 1, clusteragemonths := round(clusteragemonths, 2)]
    dpart[, clustergrowthrate := round(clustersize/clusteragemonths, 2)]
    
    # Calculate weekly growth rate over the last four weeks:
    dpart[, maxwkdate := as.Date(paste(substring(maxweek, 1, 4), substring(maxweek, 5, 6), 1, sep = "-"), "%Y-%U-%u")]
    dpart[, clustersize4wk := sum(sampleorigin == "human" & (reportweek >= as.numeric(isoyrwk(maxwkdate - weeks(4))))), by = eval(groupvector)]
    dpart[ sampleorigin == "human", fourweekgr := round(clustersize4wk/4, 2)]
    
    # Identify animal and food samples and remove them:
    dpart[ sampleorigin != "human", remove := TRUE]
    dpart = dpart[ remove == FALSE ] # Removes non-human samples
    
    # Calculate min, max and mean age:
    dpart[, agemin := min(age, na.rm = TRUE), by = eval(groupvector)]
    dpart[, agemax := max(age, na.rm = TRUE), by = eval(groupvector)]
    dpart[, agemedian := round(median(age, na.rm = TRUE), 0), by = eval(groupvector)]
    dpart[, agerange := paste(agemin, " - ", agemax, sep = "")]
    dpart[, adult := length(which(age >= 16)), by = eval(groupvector)]
    dpart[, total4age := length(which(!is.na(age))), by = eval(groupvector)]
    dpart[, adultprop := round((adult/total4age)*100, 0)]
    
    # Calculate sex, sex ratio and ID gum clinics:
    dpart[ sex != "f" & sex != "m", sex := NA]
    dpart[, female := length(which(sex == "f")), by = eval(groupvector)]
    dpart[, male := length(which(sex == "m")), by = eval(groupvector)]
    dpart[, total4sex := length(which(!is.na(sex))), by = eval(groupvector)]
    dpart[, maleprop := round((male/total4sex)*100, 0)]
    dpart[, sexratio := paste("F ", female, " : ", "M ", male, sep = "")]
    dpart[, gumcase := FALSE]
    dpart[ grepl("[[:digit:]]", firstname) == TRUE | grepl("[[:digit:]]", surname) == TRUE, gumcase := TRUE]
    dpart[, gumprop := round((length(which(gumcase == TRUE)) / clustersize), 2), by = eval(groupvector)]
    
    # Fix and calculate foreign travel:
    uk <- c("england", "wales", "northern ireland", "n ireland", "scotland", "isle of man")
    dnations <- c("northern ireland", "n ireland", "scotland", "wales", "isle of man")
    
    dpart[ is.na(traveldestination) & country %!in% uk, traveldestination := country]
    dpart[ !is.na(traveldestination), foreigntravel := "yes"]
    dpart[ foreigntravel == "y", foreigntravel := "yes"]
    dpart[ is.na(traveldestination) & foreigntravel == "n", foreigntravel := "no"]
    dpart[ is.na(foreigntravel), foreigntravel := "unknown"]
    
    dpart[ is.na(traveldestination) & foreigntravel == "yes", traveldestination := "unknown destination"]
    dpart[ is.na(traveldestination) & foreigntravel == "no", traveldestination := "domestic"]
    dpart[ is.na(traveldestination) & foreigntravel == "unknown", traveldestination := "no travel history"]
    
    # Temporary fix for country code errors:
    dpart[ traveldestination == "macau", traveldestination := "morocco"]
    dpart[ traveldestination == "panama", traveldestination := "pakistan"]
    
    # Calculate main travel destination by cluster:
    dpart[, samedestination := .N, by = c((groupvector), "traveldestination")]
    dpart[, maxdestination := max(samedestination, 0, na.rm = T), by = eval(groupvector)]
    dpart[, maindestination := traveldestination[which.max(samedestination)], by = eval(groupvector) ]
    dpart[, maindestinationprop := round((maxdestination / clustersize)*100, 0), by = eval(groupvector)]
    
    # New column 'any travel' to identify clusters with any travel destination even if proportion not high enough to mark them as travel
    dpart[, anytravel := paste0(unique(traveldestination[traveldestination %!in% c("unknown destination", "domestic", "no travel history")]), collapse = " "), by = eval(groupvector)]
    
    ## Create devolved nation columns:
    dpart[, wales := ifelse(any(country == "wales", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, scotland := ifelse(any(country == "scotland", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, nireland := ifelse(any(country == "n ireland" | country == "northern ireland", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, isleofman := ifelse(any(country == "isle of man", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    
    
    # Add PHEC columns:
    dpart[, eastmidlands := ifelse(any(phec == "east midlands", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, eastofengland := ifelse(any(phec == "east of england", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, london := ifelse(any(phec == "london", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, northeast := ifelse(any(phec == "north east", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, northwest := ifelse(any(phec == "north west", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, southeast := ifelse(any(phec == "south east", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, southwest := ifelse(any(phec == "south west", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, westmidlands := ifelse(any(phec == "west midlands", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, yandh := ifelse(any(phec == "yorkshire and humber", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    dpart[, seal := ifelse(any(phec == "london" | phec == "south east", na.rm = TRUE), TRUE, FALSE), by = eval(groupvector)]
    
    # Create FES column:
    dpart[, fes := phec]
    dpart[ phec == "london" | phec == "south east", fes := "south east and london"]
    
    # Create geographic filter columns:
    dpart[, phecs := nlevels(as.factor(phec)), by = eval(groupvector)]
    dpart[, festeams := nlevels(as.factor(fes)), by = eval(groupvector)]
    dpart[, devnations := nlevels(as.factor(country[country %in% dnations])), by = eval(groupvector)]
    dpart[, othercountries := nlevels(as.factor(country[country %!in% uk])), by = eval(groupvector)]
    dpart[, noregions := phecs + devnations + othercountries, by = eval(groupvector)]
    dpart[ noregions <= 1, geography := "single centre"]
    dpart[ noregions > 1, geography := "multicentre"]
    dpart[ geography == "single centre" & phecs == 1, location := phec]
    dpart[ geography == "multicentre" & noregions == 2 & phecs == 2 & festeams == 1 & seal == TRUE, location := fes]
    dpart[ geography == "multicentre" & noregions == 2 & phecs == 2 & festeams == 1 & seal == TRUE, geography := "single centre"]
    dpart[ geography == "single centre" & (devnations == 1 | othercountries == 1), location := country]
    dpart[ geography == "multicentre", location := paste(phecs, "PHECs &", devnations, "DAs &", othercountries, "other countries")]
    
    dpart[ othercountries >= 1 & othercountries <= 3, international := paste0(unique(country), collapse = " "), by = eval(groupvector)]
    dpart[ othercountries > 3, international := paste(othercountries, "non-UK countries")]
    
    
    # Summarise number and PHEC / devolved nation of residence for new cases (current reporting week) in each 5-SNP cluster:
    dpart[, newcases := sum(reportweek >= ryw), by = eval(groupvector)]
    dpart[, residence := phec]
    dpart[ is.na(phec), residence := country]
    dpart[, newcasegeog := paste0(unique(residence[!is.na(residence) & sampleorigin == "human" & reportweek >= as.numeric(ryw)]), collapse = ", "), by = eval(groupvector)]
    
    # Calculate most popular PHEC / area of residence for each 5-SNP cluster:
    dpart[, sameresidence := .N, by = c((groupvector), "residence")]
    dpart[, maxresidence := max(sameresidence, 0, na.rm = T), by = eval(groupvector)]
    dpart[, mainresidence := residence[which.max(sameresidence)], by = eval(groupvector) ]
    dpart[, mainresidenceprop := round((maxresidence / clustersize)*100, 0), by = eval(groupvector)]
    
    # Determine if household cluster:
    dpart[, hpt := pathpt]
    dpart[ is.na(pathpt), hpt := labhpt]
    
    dpart[, houseid := patpostcode]
    dpart[ is.na(patpostcode), houseid := paste0(soundexlastname, hpt)]
    dpart[, samehouse := .N, by = c((groupvector), "houseid")]
    dpart[, maxhouse := max(samehouse, 0, na.rm = TRUE), by = eval(groupvector)]
    
    # Create transmission type (cluster category) column:
    dpart[, msm := FALSE]
    dpart[ ((clustersize <= 5 & maleprop == 100 & agemin >= 16) | (clustersize > 5 & maleprop >= 90 & adultprop >= 90)), msm := TRUE]
    
    dpart[, household := FALSE]
    dpart[ maxhouse == clustersize, household := TRUE]
    
    dpart[, travel := FALSE]
    dpart[ maindestination != "no travel history" & maindestination != "unknown destination" & maindestination != "domestic" & maindestinationprop >= 50 & maxdestination >= 2, travel := TRUE]
    
    dpart[, community := FALSE]
    dpart[ msm == FALSE & household == FALSE & travel == FALSE, community := TRUE]
    
    dpart[ community == TRUE & travel == FALSE & household == FALSE & msm == FALSE, category := "community"]
    dpart[ travel == TRUE & community == FALSE & (msm == TRUE | msm == FALSE) & (household == TRUE | household == FALSE), category := "travel"]
    dpart[ household == TRUE & community == FALSE & travel == FALSE & (msm == TRUE | msm == FALSE), category := "household"]
    dpart[ msm == TRUE & community == FALSE & travel == FALSE & household == FALSE, category := "adult male"]
    
    dpart[ category == "travel", category := paste("travel to", maindestination)]
    
    dpart[, clustertype := category]
    dpart[ grepl("travel", category), clustertype := "travel"]
    
    # Create activity status column:
    dpart[ (maxweek >= ryw & minweekhuman >= ryw) | (maxweek >= ryw & totalisolates - newcases <= 1), status := "new"]
    dpart[ maxweek >= ryw & minweekhuman < ryw & totalisolates - newcases > 1, status := "active"]
    dpart[ maxweek < ryw & minweekhuman < ryw, status := "sleeping"]
    
    # Create team responsible column:
    dpart[ geography %in% c("multicentre", "international") & category == "community", team := "GI"]
    dpart[ geography %in% c("multicentre", "international") & category == "adult male", team := "STI"]
    dpart[ travel == TRUE, team := "T&MH"]
    dpart[ geography == "single centre" & travel == FALSE, team := location]
    dpart[ geography == "single centre" & travel == FALSE & (team == "south east" | team == "london"), team := "south east and london"]
    dpart[ is.na(team), team := "GI"]
    
    # Create additional pathogen columns to deal with multiple serovar issue:
    dpart[, ReportPathogen := maxtable(pathogen), by = eval(groupvector)]
    dpart[, samepathogen := .N, by = c((groupvector), "pathogen")]
    dpart[, maxpathogen := max(samepathogen, 0, na.rm = T), by = eval(groupvector)]
    dpart[, reportpathogenprop := round((maxpathogen / clustersize)*100, 0), by = eval(groupvector)]
    dpart[, pathogenmix := uniquewords(pathogen), by = eval(groupvector)]
    
    # Add to list:
    datalist[[i]] = dpart
    
    }
  
  # Bind all the summaries together:
  clusterlog = rbindlist(l = datalist, use.names = T, fill = T, idcol = "ReportNo")
  
  # List of columns to include in the report output:
  reportcols = c(eval(groupvector), "team", "status", "pipeline", 
                  "ReportPathogen", "reportpathogenprop", "pathogenmix", "clonalcomplex", "xsnpaddress", 
                  "category", "clustersize", "clusteragemonths", "clustergrowthrate", 
                  "lastspecdate", "lastseqdate", "maxweek", "clustersize4wk", "fourweekgr", "newcases", 
                  "agerange", "agemedian", 
                  "sexratio", "maleprop", 
                  "anytravel", "maindestination", "maindestinationprop", "linkedsamples", 
                  "geography", "location", "mainresidence", "mainresidenceprop", "international", "newcasegeog", 
                  "wales", "scotland", "nireland", "isleofman", "seal", "london", "southeast", "eastmidlands", "eastofengland", "northeast", "northwest", "southwest", "westmidlands", "yandh")
  
  # Create cluster tab:
  clustertab = clusterlog[ location != "ireland", reportcols, with = F]
  
  # Set the key prior to deduplicating:
  setkeyv(clustertab, c(groupvector))
  
  # Deduplicate to get the summary table:
  clustertab = unique(clustertab)
  
  # Sort table by ID and report week:
  setorderv(clustertab, c(groupvector, "maxweek", "ReportPathogen", "clonalcomplex", "xsnpaddress", "team", "status"))
  
  # Create comments column:
  clustertab[, comments := ""]
  clustertab[ Sys.Date() - lastspecdate >= 93, comments := "Note: latest specimen > 3 months old" ]
  
  # Create linelists (unique case-level data only):
  linelistcols = c(eval(groupvector), "pathogen", "clonalcomplex", "mlst_st", "xsnpaddress", "clustersize", "category",
                    "molis_id", "ext_id", "labname", "sampletype", "sampleorigin", "sampledetails",
                    "pid", "firstname", "surname", "dob", "age", "nhsno", "sex", 
                    "foreigntravel", "traveldestination", "invasive", "gumcase",
                    "patpostcode", "geopostcode", "latitude", "longitude", "hpt", "phec", "country",
                    "snip_address", 
                    "sample_date",  "receipt_date", "reportdate_pt", "specimendate", "sequencing_date", "reportdate_seq", 
                    "isospecweek", "isorepweekseq", "isorepweekpt", "reportweek", "minweekhuman", "status", c(amrcols))
  
  clusterll = clusterlog[ location != "ireland", linelistcols, with = F]
  
  # Set the key prior to deduplicating:
  setkeyv(clusterll, c("pid"))
  
  # Deduplicate:
  clusterll = unique(clusterll)
  
  # Set row order:
  setorder(clusterll, status, pathogen, clonalcomplex, xsnpaddress, -specimendate)
  
  # Return the summarised data.tables in a list:
  dtlist <- list("linelist" = clusterll, "clustersum" = clustertab)
  return(dtlist)
  
}

#' @example 
#' # Not run:
#' snpout <- sapply(idlist, sumclustersaudit, data = snplists, simplify = FALSE, USE.NAMES = TRUE)
```

**Comment by AmyMikhail** [comment]
Update: base R crashes with the same error details:

```
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	Rgui.exe
  Application Version:	3.42.7832.0
  Application Timestamp:	59ccc2b1
  Fault Module Name:	datatable.dll
  Fault Module Version:	0.0.0.0
  Fault Module Timestamp:	5a39aedc
  Exception Code:	c0000005
  Exception Offset:	0000000000029060
  OS Version:	6.1.7601.2.1.0.256.48
  Locale ID:	2057
  Additional Information 1:	cb11
  Additional Information 2:	cb11abf51219d08bb34e1d4ff9f1a95b
  Additional Information 3:	ff0f
  Additional Information 4:	ff0f0ac722dc1c648282a37d7681e735
```

**Comment by AmyMikhail** [comment]
Update 2: `Rterm.exe` (which curiously is what opened when I clicked on `R.exe`) also crashes:

```
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	Rterm.exe
  Application Version:	3.42.7832.0
  Application Timestamp:	59ccc2b3
  Fault Module Name:	datatable.dll
  Fault Module Version:	0.0.0.0
  Fault Module Timestamp:	5a39aedc
  Exception Code:	c0000005
  Exception Offset:	0000000000029060
  OS Version:	6.1.7601.2.1.0.256.48
  Locale ID:	2057
  Additional Information 1:	56c2
  Additional Information 2:	56c2cbd44ff199678d57bf6e48a1f624
  Additional Information 3:	7c2f
  Additional Information 4:	7c2f3e180b4d57616b407bae15c5c322
```

**Comment by AmyMikhail** [comment]
Update 3: 

On further investigation, dumping the summarized files to .csv I was able to determine that the problem was not in summarizing the data, but rather with the call to `rbindlist` - where rbindlist in the CRAN released version of data.table cannot handle empty tables and crashes R as described in issue #2340 .   

The issue was fixed in data.table 1.10.5 (development version) and I'm happy to report that after upgrading to 1.10.5 my real function runs on my real data and produces the desired output without crashing R.  

Although I was unable to reproduce the problem with a MWE, I think that is just because the MWE didn't sufficiently reflect the complexity of my real data set.  I will close the issue.
